File: go/internal/client/runner.go
// Package client: CLIEntry wraps an example's main body and translates the
// typed error tree into the contract-mandated process exit codes.
package client

import (
	"context"
	"errors"
	"fmt"
	"os"
	"os/signal"
	"syscall"
)

// CLIEntry runs fn with a SIGINT/SIGTERM-aware context and calls os.Exit
// with a code that matches the language-agnostic exit policy described in
// EXAMPLES.md. SIGINT during fn execution is a clean shutdown (exit 0).
func CLIEntry(fn func(ctx context.Context) error) {
	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
	defer stop()

	err := fn(ctx)
	os.Exit(ExitCodeFor(ctx, err))
}

// ExitCodeFor converts an error returned by an example main function into
// the process exit code mandated by the shared contract. It is exported so
// tests can validate the mapping without spawning a subprocess.
func ExitCodeFor(ctx context.Context, err error) int {
	if err == nil {
		return 0
	}

	// A cancelled context (Ctrl+C) is a clean exit, not a failure.
	if ctx != nil && ctx.Err() != nil && errors.Is(err, context.Canceled) {
		return 0
	}

	var authErr *AuthError
	if errors.As(err, &authErr) {
		fmt.Fprintln(os.Stderr, err.Error())
		return 2
	}
	var rlErr *RateLimitError
	if errors.As(err, &rlErr) {
		fmt.Fprintln(os.Stderr, err.Error())
		return 3
	}
	var logicErr *ClientLogicError
	if errors.As(err, &logicErr) {
		fmt.Fprintln(os.Stderr, err.Error())
		return 4
	}
	var transientErr *TransientError
	if errors.As(err, &transientErr) {
		fmt.Fprintln(os.Stderr, err.Error())
		return 1
	}

	// Unclassified errors are treated as transient by default.
	fmt.Fprintln(os.Stderr, err.Error())
	return 1
}

File: typescript/src/runner.ts
/**
 * Shared CLI entry-point. Applies the uniform exit-code policy from
 * EXAMPLES.md ("Conventions for language implementers") to every example.
 */

import { AuthError, ClientLogicError, RateLimitError, TransientError } from "./errors.js";

export type Main = () => Promise<number | void>;

/**
 * Wrap an example's `main` function with the uniform exit-code mapping.
 *
 * Exit codes:
 *   0  success (default if `main` resolves without a number)
 *   1  TransientError (5xx / network / timeout)
 *   2  AuthError (HTTP 401)
 *   3  RateLimitError (HTTP 429)
 *   4  ClientLogicError (bad env, malformed response, etc.)
 */
export function cliEntry(main: Main): void {
  main()
    .then((code) => {
      process.exit(typeof code === "number" ? code : 0);
    })
    .catch((err: unknown) => {
      const message = err instanceof Error ? err.message : String(err);
      process.stderr.write(`error: ${message}\n`);
      if (err instanceof AuthError) process.exit(2);
      if (err instanceof RateLimitError) process.exit(3);
      if (err instanceof TransientError) process.exit(1);
      if (err instanceof ClientLogicError) process.exit(4);
      process.exit(1);
    });
}
